& ris-bali [ENH] statsmodels DynamicFactor interface

lbventura opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look
at our contribution guide: https://github.com/alan-turing-institute/sktime/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.

Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->

Addresses #2349. There are multiple failed tests due to the behavior of the `_predict` method, but further exploration in `testfile.ipynb` suggests that the method's behavior is actually correct. Could be that tests are wrong?

#### What does this implement/fix? Explain your changes.
<!--
A clear and concise description of what you have implemented.
-->

Deeper look into #2349 failed tests using check_estimator.

#### Does your contribution introduce a new dependency? If yes, which one?

<!--
If your contribution does add a new hard dependency, we may suggest to initially develop your contribution in a separate companion package in https://github.com/sktime/ to keep external dependencies of the core sktime package to a minimum.
-->

#### What should a reviewer concentrate their feedback on?

<!-- This section is particularly useful if you have a pull request that is still in development. You can guide the reviews to focus on the parts that are ready for their comments. We suggest using bullets (indicated by * or -) and filled checkboxes [x] here -->

#### Any other comments?
<!--
Please be aware that we are a loose team of volunteers so patience is necessary; assistance handling other issues is very welcome. We value all user contributions, no matter how minor they are. If we are slow to review, either the pull request needs some benchmarking, tinkering, convincing, etc. or more likely the reviewers are simply busy. In either case, we ask for your understanding during the review process.
-->

(At least) Two outstanding points:
1. Implement the `predict_proba` method;
2. Create an example notebook - one cool example would be to forecast economic variables;

#### PR checklist
<!--
Please go through the checklist below. Please feel free to remove points if they are not applicable.
-->

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/alan-turing-institute/sktime/blob/main/.all-contributorsrc).
- [ ] Optionally, I've updated sktime's [CODEOWNERS](https://github.com/alan-turing-institute/sktime/blob/main/CODEOWNERS) to receive notifications about future changes to these files.
- [ ] I've added unit tests and made sure they pass locally.
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG] indicating whether the PR topic is related to enhancement, maintenance, documentation, or bug.

##### For new estimators
- [x] I've added the estimator to the online documentation.
- [ ] I've updated the existing example notebooks or provided a new one to showcase how my estimator works.


<!--
Thanks for contributing!
-->

fkiraly commented:
I think I understand the problem now
the test is fine, it intentionally checks for indices that do not start at zero
the problem is with the (inner, statsmodels) dynamic factor model, it discards indices and produces forecasts starting at len(y) rather than at y.index[-1]
the solution would be to manually correct for that and simply overwrite the indices of what comes out of the model

lbventura commented:
Weirdly pre-commit is failing in "Install and test / code-quality (pull_request)" due to flake8, but worked in my local machine? See the latter below
<img width="613" alt="Failing_Test" src="https://user-images.githubusercontent.com/68004282/178152704-f9d6a24a-677c-4f62-8df5-3e288f5256e7.png">

fkiraly commented:
You probably don't have all the linting checks installed locally then?
You can always go on "details" here to see what's wrong:
![image](https://user-images.githubusercontent.com/7985502/178153979-cf9a18a2-4e96-47f7-b8cc-f1966f900552.png)

(I know what's wrong, and it is easy to fix - but kindly have a look yourself 😄 )

fkiraly commented on sktime/forecasting/dynamic_factor.py:
```diff
@@ -0,0 +1,429 @@
+# -*- coding: utf-8 -*-
+# copyright: sktime developers, BSD-3-Clause License (see LICENSE file)
+
+"""Implements DynamicFactor Model."""
```
should say sth like "interface to statemodels etc", like in the class docstring

fkiraly commented on sktime/forecasting/dynamic_factor.py:
```diff
@@ -0,0 +1,429 @@
+    """
+
+    _tags = {
+        "scitype:y": "multivariate",
```
this means that the data *must* have 2 or more variables. Does it not also work for 1 variable? Then the tag should be `"both"`.

fkiraly commented on sktime/forecasting/dynamic_factor.py:
```diff
@@ -0,0 +1,429 @@
+        "requires-fh-in-fit": False,
+        "X-y-must-have-same-index": True,
+        "enforce_index_type": None,
+        "capability:pred_int": False,
```
hm, I thought, based on the formulation, that the model should be able to make proba predictions?

Though, I cannot find a method for that, so maybe let's leave this to another PR but make a note in the issue?

fkiraly commented on sktime/forecasting/dynamic_factor.py:
```diff
@@ -0,0 +1,429 @@
+        cov_kwds=None,
+        method="lbfgs",
+        maxiter=50,
+        full_output=False,  # Changed from 1
```
what does this mean?

lbventura replied:
I disabled the output when I was conducting testing because of "verboseness" but will reset to the defaults.

fkiraly commented on sktime/forecasting/dynamic_factor.py:
```diff
@@ -0,0 +1,429 @@
+from sktime.utils.validation._dependencies import _check_soft_dependencies
+
+_all_ = ["DynamicFactor"]
+__author__ = ["Ris-Bali"]
```
please add yourself as an author, too.

fkiraly replied:
and in allcontributors, if you haven't done so (if you want to take the credit 😄). If @Ris-Bali is not there, we should also add them for the partial work.

fkiraly reviewed (changes requested):
Looks good!

Blocking comments:
* we need to remove the notebook `testfile.ipynb`.
* the estimator should be added to the API reference in an appropriate section.

I also left a number of small, non-blocking comments, above.

fkiraly commented on sktime/forecasting/dynamic_factor.py:
```diff
@@ -0,0 +1,429 @@
+_all_ = ["DynamicFactor"]
+__author__ = ["Ris-Bali"]
+
+_check_soft_dependencies("dynamicfactor", severity="warning")
```
there is no such package, at least it is not imported. Also, `statsmodels` is a core dependency, so you can remove this line.

lbventura commented:
> the estimator should be added to the API reference in an appropriate section.



> Looks good!
> 
> Blocking comments:
> 
>     * we need to remove the notebook `testfile.ipynb`.
> 
>     * the estimator should be added to the API reference in an appropriate section.
> 
> 
> I also left a number of small, non-blocking comments, above.

On the second point, what exactly should be done? One example would be helpful :)

fkiraly commented:
> On the second point, what exactly should be done? One example would be helpful :)

goto `docs.source.api_reference.forecasting` and add an entry, following the pattern.
If not sure, jump quickly on discord.

fkiraly commented:
... odd, looks correct.

fkiraly commented:
ah, looks like a typo in the "references" section of the new estimator. See "details" next to the doc build, maybe you can find out what exactly the typo is.

lbventura commented:
It is complaining that the output is too long? Could it be due to the fact that I changed `full_output` to `True`?

> ... odd, looks correct.

It is complaining that the output is too long? Could it be due to the fact that I changed `full_output` to `True`?

fkiraly commented:
hm, estimator tests are failing now - any idea why? Perhaps you would like to change it back to the last state where they were not?

lbventura commented:
> hm, estimator tests are failing now - any idea why? Perhaps you would like to change it back to the last state where they were not?

The only reasons I can think of are:

1.  Having changed the sciptype from multivariate to both. I reverted back to multivariate;
2. Having changed the outputs of the model (the `full_output` point discussed above).

Let us see if the new commit addresses the problem.

fkiraly commented:
>The only reasons I can think of are:

>1.  Having changed the sciptype from multivariate to both. I reverted back to multivariate;

But, should it not be `both`? This estimator can deal with univariate data, can it not?

> 2. Having changed the outputs of the model (the `full_output` point discussed above).

Why did you do this?

The docs are still not building, btw, kindly check for typos.

lbventura commented:
> > The only reasons I can think of are:
> 
> > 1. Having changed the sciptype from multivariate to both. I reverted back to multivariate;
> 
> But, should it not be `both`? This estimator can deal with univariate data, can it not?

This was just a quick edit to see whether the problem comes from here. I think it should also work, but I will check.
> 
> > 2. Having changed the outputs of the model (the `full_output` point discussed above).
> 
> Why did you do this?

The docs were being built fine in a previous commit, reverted to see if this was the change that broke it, but aparently not.

> The docs are still not building, btw, kindly check for typos.

Will do

fkiraly commented:
> I think it should also work for univariate, but I will check. EDIT: it also does not work in statsmodels. `y` is just a 1D series generated with `_make_series`.

Ah, ok - then the value `"multivariate"` that you originally had was correct.

lbventura commented:
Tried to replicate `arima.py` forecaster docstring, hopefully this solves the readthedocs problem. Unfortunately the build logs are not very clear.

fkiraly commented:
> hopefully this solves the readthedocs problem. Unfortunately the build logs are not very clear.

Yes, looks like it! All green!

fkiraly reviewed (approved):
Nice, thanks!

lbventura commented:
> Nice, thanks!

Great, on to the next topic!
